File: src/lib/server/desk.ts
import { SKILLS } from "../catalog";
import { findBlock, parseTradeIntent } from "../memory/ingest";
import type { IngestResult, MemorySnapshot } from "../memory/types";
import type { Market, PortfolioSnapshot, ProposedTrade } from "../types";

/** Deterministic desk — not an LLM. Runs the loaded skill against the tape and the book memory. */
export function deskBrief(input: {
  text: string;
  skills: string[];
  book: PortfolioSnapshot;
  markets: Market[];
  focus?: string;
  memory?: MemorySnapshot;
  ingest?: IngestResult;
}): { text: string; trades: ProposedTrade[] } {
  const focus = input.focus ?? "BTC";
  const market = input.markets.find((m) => m.symbol === focus) ?? input.markets[0];
  const skill = SKILLS.find((s) => input.skills.includes(s.id)) ?? SKILLS[0]!;
  const heat = input.book.positions.reduce((a, p) => a + Math.abs(p.qty) * p.avgPrice, 0);
  const up = (market?.change24h ?? 0) >= 0;
  const wantTrade = /trade|buy|sell|size|probe|enter|cut|short|long/i.test(input.text);
  const trades: ProposedTrade[] = [];
  const intent = parseTradeIntent(input.text, input.markets, focus);
  const block = input.memory ? findBlock(input.memory, intent) : null;

  let action = "Stand aside.";
  let reason = `${skill.name}: no confirmation worth paying for.`;

  if (input.ingest?.wrote) {
    action = "Recorded.";
    reason = input.ingest.summary ?? "Wrote that into book memory. It survives this chat.";
  } else if (block) {
    action = "Refuse.";
    reason = `Memory ${block.entityName}: ${block.reason}`;
  } else if (market && skill.id === "turtle" && Math.abs(market.change24h) > 1.2) {
    action = up ? `Probe long ${market.symbol}.` : `Probe short ${market.symbol} (perp only).`;
    reason = `Channel-style expansion (${market.change24h.toFixed(2)}%). Mechanical unit, not a story.`;
  } else if (market && (skill.id === "livermore" || skill.id === "minervini-vcp") && market.change24h > 1) {
    action = `Probe long ${market.symbol}.`;
    reason = `Line of least resistance still up. Small first unit; pyramid only if it pays.`;
  } else if (market && (skill.id === "druckenmiller" || skill.id === "hayes-liquidity") && up) {
    action = `Keep risk on via ${market.symbol}, do not chase.`;
    reason = `Liquidity-style tape is still bid. Size from cash, not from boredom.`;
  } else if (market && skill.id === "buffett") {
    action = "Cash is a position.";
    reason = "Owner-earnings cadence. A 24h print is not a thesis.";
  } else if (market && market.change24h < -2) {
    action = "Reduce heat. Do not average a break.";
    reason = `${skill.name} risk rule: a failed tape is an exit, not a debate.`;
  }

  if (block && !(input.ingest?.wrote && !/\bagain\b/i.test(input.text))) {
    action = "Refuse.";
    reason = `Memory ${block.entityName}: ${block.reason}`;
  }

  if (wantTrade && market && !input.book.killSwitch && action.startsWith("Probe") && !block) {
    const px = market.price;
    const budget = Math.min(input.book.cash * 0.05, 2500);
    const qty = Number((budget / px).toPrecision(3));
    if (qty > 0) {
      trades.push({
        symbol: market.venue === "perp" && !up ? market.symbol : market.symbol,
        side: up ? "buy" : market.venue === "perp" ? "sell" : "buy",
        type: "market",
        qty,
        leverage: market.venue === "perp" ? 2 : 1,
        reason,
      });
    }
  }

  const pos =
    input.book.positions.length === 0
      ? "flat"
      : input.book.positions.map((p) => `${p.side} ${p.qty} ${p.symbol}`).join(", ");

  const lessons = (input.memory?.entities ?? []).filter((e) => e.status === "live" && e.category === "lesson");

  const text = [
    `${skill.author} / ${skill.name}`,
    "",
    market
      ? `${market.symbol}  ${market.price.toPrecision(6)}  ${market.change24h >= 0 ? "+" : ""}${market.change24h.toFixed(2)}%  (${market.venue})`
      : "No mark.",
    `Book  cash ${input.book.cash.toFixed(0)}  equity ${input.book.equity.toFixed(0)}  heat ${heat.toFixed(0)}  ${pos}`,
    lessons.length ? `Memory  ${lessons.map((e) => e.name).join(" · ")}` : "Memory  empty",
    "",
    action,
    reason,
    "",
    `Playbook: ${skill.playbooks[0]}`,
    `Risk: ${skill.risk}`,
    input.book.killSwitch ? "Kill switch is on — no execution." : "",
    trades.length ? `Ticket queued: ${trades.map((t) => `${t.side} ${t.qty} ${t.symbol}`).join("; ")}` : "",
    "",
    "Desk Rules is local and systematic. It is not Grok, and it is not advice.",
  ]
    .filter(Boolean)
    .join("\n");

  return { text, trades };
}

File: desktop/src/desk.rs
use crate::book::Book;
use crate::memory::{find_block, ingest, Memory};
use crate::okx::Ticker;

pub struct Brief {
    pub text: String,
    pub side: Option<String>,
    pub qty: Option<f64>,
}

pub fn brief(text: &str, focus: &str, ticker: Option<&Ticker>, book: &Book, memory: &mut Memory) -> Brief {
    let wrote = ingest(memory, text, focus);
    let up = ticker.map(|t| t.change24h >= 0.0).unwrap_or(true);
    let want = regex_want(text);
    let intent_side = if text.to_lowercase().contains("short") {
        "short"
    } else if text.to_lowercase().contains("buy") || text.to_lowercase().contains("long") {
        "long"
    } else if want {
        if up { "long" } else { "short" }
    } else {
        ""
    };
    let symbol = pick_symbol(text, focus);
    let block = if intent_side.is_empty() {
        None
    } else {
        find_block(memory, &symbol, intent_side)
    };

    let mut action = "Stand aside.".to_string();
    let mut reason = "No confirmation worth paying for.".to_string();
    let mut side = None;
    let mut qty = None;

    if wrote {
        action = "Recorded.".into();
        reason = "Wrote that into book memory. It survives this chat.".into();
    } else if let Some(b) = &block {
        action = "Refuse.".into();
        reason = format!("Memory {b}");
    } else if let Some(t) = ticker {
        if t.change24h.abs() > 1.2 {
            action = if up {
                format!("Probe long {}.", t.symbol)
            } else {
                format!("Probe short {}.", t.symbol)
            };
            reason = format!("Tape expansion {:.2}%. Mechanical unit.", t.change24h);
        }
    }

    if let Some(b) = &block {
        if !wrote || text.to_lowercase().contains("again") {
            action = "Refuse.".into();
            reason = format!("Memory {b}");
        }
    }

    if want && action.starts_with("Probe") && !book.kill {
        if let Some(t) = ticker {
            let budget = (book.cash * 0.05).min(2500.0);
            let q = ((budget / t.last) * 1000.0).round() / 1000.0;
            if q > 0.0 {
                side = Some(if up { "buy".into() } else { "sell".into() });
                qty = Some(q);
            }
        }
    }

    let pos = if book.positions.is_empty() {
        "flat".into()
    } else {
        book.positions
            .iter()
            .map(|p| format!("{} {} {}", p.side, p.qty, p.symbol))
            .collect::<Vec<_>>()
            .join(", ")
    };
    let lessons: Vec<_> = memory
        .lessons
        .iter()
        .filter(|l| l.live)
        .map(|l| l.name.clone())
        .collect();

    let mark = ticker
        .map(|t| format!("{}  {:.4}  {:+.2}%", t.symbol, t.last, t.change24h))
        .unwrap_or_else(|| "No mark.".into());

    let text = format!(
        "Desk Rules / Liquidity Macro\n{mark}\nBook  cash {:.0}  {pos}\nMemory  {}\n\n{action}\n{reason}\n\nDesk Rules is local and systematic. It is not advice.",
        book.cash,
        if lessons.is_empty() {
            "empty".into()
        } else {
            lessons.join(" · ")
        }
    );

    Brief { text, side, qty }
}

fn regex_want(text: &str) -> bool {
    let t = text.to_lowercase();
    ["trade", "buy", "sell", "probe", "enter", "short", "long", "size"]
        .iter()
        .any(|w| t.contains(w))
}

fn pick_symbol(text: &str, focus: &str) -> String {
    let upper = text.to_uppercase();
    for s in crate::okx::UNIVERSE {
        if upper.split(|c: char| !c.is_ascii_alphanumeric()).any(|tok| tok == *s) {
            return (*s).into();
        }
    }
    focus.to_uppercase()
}
